File: mvvmsmart/src/main/java/com/wzq/mvvmsmart/event/StateLiveData.java
package com.wzq.mvvmsmart.event;

/**
 * <p>作者：王志强<p>
 * <p>创建时间：2019/12/21<p>
 * <p>文件描述：带状态的livedata<p>
 */

import androidx.lifecycle.MutableLiveData;

public class StateLiveData<T> extends MutableLiveData<T> {
    public enum StateEnum {
        Idle,
        Loading,
        Success,
        Error,
        NoData,
        NoMoreData,
        NoNet
    }

    // 封装的枚举,用MutableLiveData可以被view观察;
    public MutableLiveData<StateEnum> stateEnumMutableLiveData = new MutableLiveData<>();

    public final void postValueAndSuccess(T t) {
        super.postValue(t);
        this.postSuccess();
    }

    public void postIdle() {
        stateEnumMutableLiveData.postValue(StateEnum.Idle);
    }

    public void postLoading() {
        stateEnumMutableLiveData.postValue(StateEnum.Loading);
    }

    public void postNoData() {
        stateEnumMutableLiveData.postValue(StateEnum.NoData);
    }

    public void postNoMoreData() {
        stateEnumMutableLiveData.postValue(StateEnum.NoMoreData);
    }

    public void postNoNet() {
        stateEnumMutableLiveData.postValue(StateEnum.NoNet);
    }

    public void postSuccess() {
        stateEnumMutableLiveData.postValue(StateEnum.Success);
    }

    public void postError() {
        stateEnumMutableLiveData.postValue(StateEnum.Error);
    }

    public void changeState(StateEnum stateEnum) {
        stateEnumMutableLiveData.postValue(stateEnum);
    }


}

File: mvvmsmart/src/main/java/com/wzq/mvvmsmart/event/StateLiveData.kt
package com.wzq.mvvmsmart.event

import androidx.lifecycle.MutableLiveData

/**
 *
 * 作者：王志强
 *
 *
 *
 * 创建时间：2019/12/21
 *
 *
 *
 * 文件描述：带状态的livedata
 *
 *
 */
class StateLiveData<T> : MutableLiveData<T>() {
    enum class StateEnum {
        Idle, Loading, Success, Error, NoData, NoMoreData, NoNet
    }

    // 封装的枚举,用MutableLiveData可以被view观察;
    var stateEnumMutableLiveData = MutableLiveData<StateEnum>()
    fun postValueAndSuccess(t: T) {
        super.postValue(t)
        postSuccess()
    }

    fun postIdle() {
        stateEnumMutableLiveData.postValue(StateEnum.Idle)
    }

    fun postLoading() {
        stateEnumMutableLiveData.postValue(StateEnum.Loading)
    }

    fun postNoData() {
        stateEnumMutableLiveData.postValue(StateEnum.NoData)
    }

    fun postNoMoreData() {
        stateEnumMutableLiveData.postValue(StateEnum.NoMoreData)
    }

    fun postNoNet() {
        stateEnumMutableLiveData.postValue(StateEnum.NoNet)
    }

    fun postSuccess() {
        stateEnumMutableLiveData.postValue(StateEnum.Success)
    }

    fun postError() {
        stateEnumMutableLiveData.postValue(StateEnum.Error)
    }

    fun changeState(stateEnum: StateEnum) {
        stateEnumMutableLiveData.postValue(stateEnum)
    }
}